jubilant/_helpers.py

```diff
@@ -0,0 +1,18 @@
+from ._types import Status
+
+
+def any_error(status: Status) -> bool:
+    """Return true if any application or unit in *status* is in "error" status."""
+    for app_info in status.applications.values():
+        if app_info.application_status.current == 'error':
+            return True
+        for unit_info in app_info.units.values():
+            if unit_info.workload_status.current == 'error':
+                return True
+        # TODO: what about juju_status (agent status)?
+    return False
+
+
+def any_blocked(status: Status) -> bool:
+    """Return true if any application or unit in *status* is in "blocked" status."""
+    return False  # TODO
```

Comment:
My thinking is that I'd add all the variants, at least the useful ones (probably not `any_unknown`), including `any_waiting`. Similar for `Status.is_app_[active|waiting|...]`. I like this better than `jubilant.any_status_is('waiting')` because it's too easy to make a typo in `waiting`. And if you go the `jubilant.any_status_is(jubilant.Status.WAITING)` it's too verbose.

Regarding the `*_func` variant that returns a function. Apart from being harder to compose, I find that thinking about these functions-that-return-functions is a notch harder, like a [dream within a dream](https://www.imdb.com/title/tt1375666/). I'll reply with more further down where you list the examples of composing those.